cytodynamics-test-container/src/test/java/com/linkedin/cytodynamics/test/TestDynamicLoad.java

```diff
@@ -245,6 +274,124 @@ public void testLoadAandB() {
+    assertEquals(clazz.getClassLoader(), parentClassLoaderB);
+  }
+
+  /**
+   * Graph structure:
+   *
+   * (common parent  <------------ (partial delegation
+   *  classloader)                  classloader)
+   *          ^                     ^
+   *          |                    /
+   *          |                   /
+   *         FULL               NONE
+   *          |                 /
+   *          |                /
+   *         (child classloader)
+   *
+   * A more concrete use case for this is that the common parent would contain some shared API interfaces that need to
+   * be used by both the partial delegation classloader and the child classloader. The partial delegation classloader
+   * and the child classloader could be managed by different owners, so they would need to be separate and isolated.
+   *
+   * It should not be possible to build a cyclic graph with the cytodynamics builder APIs, so we are just considering an
+   * acyclic graph here.
+   */
+  @Test(description = "Given a structure in which the parent relationships form a graph, and two of the loaders have "
+      + "a common parent, then delegation should properly load from the correct loaders")
+  public void testGraphRelationshipWithCommonParent() throws Exception {
+    URL commonParentJarUrl = getTestJarUri("api").toURL();
+    ClassLoader commonParent = new URLClassLoader(new URL[]{commonParentJarUrl}, null);
+    ClassLoader partialDelegation = new URLClassLoader(new URL[]{getTestJarUri("a").toURL()}, commonParent);
+    Loader loader = LoaderBuilder
+        .anIsolatingLoader()
+        .withOriginRestriction(OriginRestriction.allowByDefault())
+        .withClasspath(Collections.singletonList(getTestJarUri("b")))
+        .addParentRelationship(ParentRelationshipBuilder.builder()
+            .withParentClassLoader(commonParent)
+            .withIsolationLevel(IsolationLevel.FULL)
+            .addParentPreferredClassPattern("java.*")
+            // TODO fix: when loader's classloader doesn't match the parent classloader, Api annotation doesn't apply
+            .addParentPreferredClassPattern(TestInterface.class.getName())
+            .build())
+        .addParentRelationship(ParentRelationshipBuilder.builder()
+            .withParentClassLoader(partialDelegation)
+            .withIsolationLevel(IsolationLevel.NONE)
+            .build())
+        .build();
+
+    /*
+     * If we tried to assign to a TestInterface variable directly, it would not work, because the TestInterface for the
+     * actual objects are loaded from the commonParent and not the classloader which is associated with the execution of
+     * this test.
+     * This means we need to use reflection to call methods on the objects.
+     */
+    Object testInterfaceImpl = loader.newInstanceOf(Object.class, TestInterfaceImpl.class.getName());
+    // check that the TestInterfaceImpl comes from the loader classpath
```

Comment:
These comments are super useful, but I would suggest inserting them in the third param of the assertion instead. This way, if the test fails, it provides a meaningful failure message instead of just saying something opaque like "expected B, but got A-only" or something like that.